qiskit_ode/signals/signals.py

```diff
@@ -231,24 +277,75 @@ def __init__(
             dt: The duration of each sample.
             samples: The array of samples.
             start_time: The time at which the signal starts.
-            duration: The duration of the signal in samples.
             carrier_freq: The frequency of the carrier.
             phase: The phase of the carrier.
             name: name of the signal.
         """
-        super().__init__(name)
-
+        self._name = name
+        self._is_constant = False
         self._dt = dt
+        self._samples = Array(samples)
+        self._start_time = start_time
 
-        if samples is not None:
-            self._samples = Array(samples)
-        else:
-            self._samples = Array([0.0] * duration)
+        # initialize internally stored carrier/phase information
+        self._carrier_freq = None
+        self._phase = None
+        self._carrier_arg = None
+        self._phase_arg = None
```

Comment:
Could this not be done in the base class? It feels strange to not call `super().__init__` here but I see why you did it since `Signal.__init__` wants to define a `self._envelope` which `DiscreteSignal` does not have. Can we not bring these classes closer together? E.g. by renaming `self._samples` to `self._envelope` in `DiscreteSignal` and letting `Signal.__init__` accept an `Array` as well as `Callable, complex, float, int` (you already do a conversion in the `__init__` of signal)

Reply:
I agree in general with this comment, though I've tried a different approach than your suggestion: 

- I've instead changed `DiscreteSignal.__init__` to construct the envelope evaluation function (basically just the code that was in `DiscreteSignal.envelope`), then call `Signal.__init__` with this envelope. 
- Following this, I changed `DiscreteSignalSum.__init__` to call `DiscreteSignal.__init__`, as the envelope evaluation functions are the same in both of these cases.
- I also changed `SignalSum.__init__` to construct the envelope function (again, from code that was originally in `SignalSum.envelope`), then again called `Signal.__init__` with this envelope (as in `DiscreteSignal`).
- As a result of the above changes I was able to eliminate:
   - Several redundant lines of code in various `__init__` functions   
   - Overriding `envelope` methods in `DiscreteSignal` and `SignalSum`.
   - Unrelated to inheritance changes, but I realized the `complex_value` method in `DiscreteSignalSum` was redundant, as it inherits the correct form of this method from `SignalSum`, so this was also removed.

I think this might clean things up now - everything calls back to the original `Signal.__init__`, with array evaluation logic set up in the respective inits.